pallets/xcm-core-buyer/src/lib.rs

```diff
@@ -0,0 +1,435 @@
+            if parathread_params.is_none() {
+                return Err(Error::<T>::NotAParathread.into());
+            }
+
+            // TODO: also compare the latest slot from pallet_author_noting with parathread_params.slot_frequency
+
+            let xcm_weights_storage =
+                XcmWeights::<T>::get().ok_or(Error::<T>::XcmWeightStorageNotSet)?;
+
+            let withdraw_amount = xcm_weights_storage.buy_execution_cost;
+
+            // Send xcm to the relay
+            // Use a derivative account from the sovereign account based on the paraId
+            // Buy on-demand cores
+            // Any failure should return everything to the derivative account
+
+            // Don't use utility::as_derivative because that will make the tanssi sovereign account
+            // pay for fees, instead use `DescendOrigin` to make the container chain sovereign account
+            // pay for fees. The container chain sovereign account is derived from the tanssi sovereign
+            // account.
+            // TODO: when coretime is implemented, buy core there instead of buying on-demand cores
+            let origin = OriginKind::SovereignAccount;
+            // TODO: max_amount is the max price of a core that this parathread is willing to pay
+            // It should be defined in a storage item somewhere, controllable by the container chain
+            // manager.
+            let max_amount = u128::MAX;
+            let call = T::GetPurchaseCoreCall::get_encoded(max_amount, para_id);
+            let weight_at_most = xcm_weights_storage.weight_at_most;
+
+            // Assumption: derived account already has DOT
+            // The balance should be enough to cover
+            // TODO: we could make this be part of the proof, so collators cannot call this if the
+            // derived account does not have enough balance
+            // Although that would not be perfect, the relay state can change in the following block,
+            // and the xcm message will be executed in the block n+2, where n is the latest relay
+            // block number seen from the tanssi block that included this extrinsic.
+            let relay_asset_total: MultiAsset = (Here, withdraw_amount).into();
+            let refund_asset_filter: MultiAssetFilter =
+                MultiAssetFilter::Wild(WildMultiAsset::AllCounted(1));
+            // TODO: need better names for this methods.
+            //  interior_multilocation is the one used in DescendOrigin
+            //  absolute_multilocation is the one used in DepositAsset
+            // They can be easily converted from one another, the difference is that absolute_multilocation
+            // has an extra "Parachain" junction in the front, using SelfParaId::get()
+            let interior_multilocation = Self::interior_multilocation(para_id);
+            let derived_account = Self::absolute_multilocation(interior_multilocation);
```

Comment:
There should be an easier way to do this, I think there is a location inverter at some point that we could incorporate

Reply:
it's the most idiomatic way to do it (it is meant for that). Comments should make sure we can understand it